### ISSUE TITLE
Equivalent of numpy.where or numpy.nonzero

### ISSUE THREAD

**Issue by tsoernes** [bluss/ndarray#466]
Is there an efficient way to get the indecies of non-zero or true elements?

**Comment by tsoernes** [comment]
The use is (partially) to select a non-zero entry at random and set it to zero, as part of a simulator. `where` and `nonzero` are pretty commonly used functions in numpy, it would be nice with a fast, built in function. Thanks anyhow

**Comment by jturner314** [comment]
Yeah, I'd like to add something like `where` or `nonzero` to `ndarray`. I was curious about your use-case because `where` and `nonzero` didn't seem very useful without support for slicing with index arrays (which I'd also like to add to `ndarray`), but using them for random selection make sense.

**Comment by bluss** [comment]
I wonder if this problem would be more efficient if we had something like a bitmap. Storing indices just sounds so inefficient and an ndimensional bitmap could be faster (just a wild idea?) way to represent a set of locations in an array.

**Comment by tsoernes** [comment]
I have also found that masks can often substitute for lists of indexes. But there are certainly use cases where this is not so. For example, I often have the need to get the values at a set of non-continuous indexes in a large matrix. If the matrix is large compared to the number of values you want to extract; e.g. you don't want to multiply a million zeroes to leave only 10 entries as-is. See e.g. the latter half of this file: https://github.com/tsoernes/rustdca/blob/master/src/gridfuncs.rs where any reference to `neigh(bor)` or `ch(annel)` is an array of indexes. Most of these algorithms cannot be feasibly be converted to the use of bitmasks. `np.where` is used all over in machine learning.

**Comment by bluss** [comment]
So, a sparse bitmask maybe?

**Comment by tsoernes** [comment]
Wouldnt a sparse bitmask be implemented by storing a list of indecies?

**Comment by bluss** [comment]
I guess a good bitmask can handle both scenarios. There's an alternative to list of indices here https://github.com/brettwooldridge/SparseBitSet/blob/master/SparseBitSet.pdf  and then there's roaring, and I'm not sure exactly what data structure it is using. https://github.com/RoaringBitmap/RoaringBitmap
